File: study/6주차 문제/장원석/정수를나선형으로배치하기.java
import java.util.Arrays;

public class 정수를나선형으로배치하기 {

	public static void main(String[] args) {
		System.out.println(Arrays.deepToString(solution(4)));
		System.out.println(Arrays.deepToString(solution(5)));
	}

	private static int[][] solution(int n) {
		int[][] answer = new int[n][n];
		int[] dx = { 0, 1, 0, -1 };
		int[] dy = { 1, 0, -1, 0 };
		int x = 0;
		int y = 0;
		int d = 0;

		for (int i = 1; i <= n * n; i++) {
			answer[x][y] = i;
			int nx = x + dx[d];
			int ny = y + dy[d];

			if (nx < 0 || ny < 0 || nx >= n || ny >= n || answer[nx][ny] != 0) {
				d++;
				if (d > 3)
					d = 0;

				nx = x + dx[d];
				ny = y + dy[d];
			}

			x = nx;
			y = ny;
		}

		return answer;
	}
}

File: study/6주차 문제/박정환/정수를 나선형으로 배치하기.py
def solution(n):
    answer = [[0] * n for _ in range(n)]
    x, y = 0, 0
    dx = [0, 1, 0, -1]
    dy = [1, 0, -1, 0]
    x2 = 2
    answer[x][y] = 1
    while x2 <= n**2:
        for i in range(len(dx)):
            while True:
                nx = x + dx[i]
                ny = y + dy[i]
                if nx < 0 or ny < 0 or nx >= n or ny >= n or answer[nx][ny] != 0:
                    break
                else:
                    answer[nx][ny] = x2
                    x = nx
                    y = ny
                    x2 += 1
    return answer
